File: app/src/main/java/com/lqk/videolearn/ui/FramePlayerActivity.java
package com.lqk.videolearn.ui;

import android.app.Activity;
import android.graphics.Matrix;
import android.graphics.SurfaceTexture;
import android.os.Bundle;
import android.util.Log;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;

import androidx.annotation.Nullable;

import com.lqk.videolearn.R;
import com.lqk.videolearn.grafika.FramePlayer;

import java.io.File;
import java.io.IOException;

/**
 * 帧播放器样例
 * @author lqk
 */
public class FramePlayerActivity extends Activity implements View.OnClickListener {
    private static final String TAG = FramePlayer.class.getSimpleName();

    /**
     * 用来显示解码结果的 Surface
     */
    TextureView mTextureView;

    /**
     * 播放线程
     */
    FramePlayer.PlayTask mPlayTask;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_player_frame);
        mTextureView = findViewById(R.id.frame_player);
        findViewById(R.id.play).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.play:
                clickPlayStop();
                break;
            default:
                break;
        }
    }

    /**
     * 播放与暂停按钮
     */
    private void clickPlayStop(){
        if (mPlayTask != null){
            return;
        }

        SurfaceTexture st = mTextureView.getSurfaceTexture();
        Surface surface = new Surface(st);

        FramePlayer framePlayer = null;
        try{
            framePlayer = new FramePlayer(new File("sdcard/ffmpegtest/test.mp4"), surface);
        }catch (IOException e){
            surface.release();
            return;
        }
        adjustAspectRatio(framePlayer.getWidth(), framePlayer.getHeight());
        mPlayTask = new FramePlayer.PlayTask(framePlayer);
        mPlayTask.execute();
    }

    /**
     * 调整TextureView的显示矩阵
     * @param videoWidth
     * @param videoHeight
     */
    private void adjustAspectRatio(int videoWidth, int videoHeight){
        int viewWidth = mTextureView.getWidth();
        int viewHeight = mTextureView.getHeight();
        double aspectRatio = (double) videoHeight / videoWidth;

        int newWidth, newHeight;
        if (viewHeight > (int) (viewWidth * aspectRatio)) {
            // limited by narrow width; restrict height
            newWidth = viewWidth;
            newHeight = (int) (viewWidth * aspectRatio);
        } else {
            // limited by short height; restrict width
            newWidth = (int) (viewHeight / aspectRatio);
            newHeight = viewHeight;
        }
        int xoff = (viewWidth - newWidth) / 2;
        int yoff = (viewHeight - newHeight) / 2;
        Log.v(TAG, "video=" + videoWidth + "x" + videoHeight +
                " view=" + viewWidth + "x" + viewHeight +
                " newView=" + newWidth + "x" + newHeight +
                " off=" + xoff + "," + yoff);

        Matrix txform = new Matrix();
        mTextureView.getTransform(txform);
        txform.setScale((float) newWidth / viewWidth, (float) newHeight / viewHeight);
        //txform.postRotate(10);          // just for fun
        txform.postTranslate(xoff, yoff);
        mTextureView.setTransform(txform);
    }
}

File: app/src/main/java/com/lqk/videolearn/ui/FramePlayerActivity.kt
package com.lqk.videolearn.ui

import android.app.Activity
import android.graphics.Matrix
import android.graphics.SurfaceTexture
import android.os.Bundle
import android.view.Surface
import android.view.TextureView
import android.view.View
import android.widget.Button
import android.widget.TextView
import com.lqk.videolearn.R
import com.lqk.videolearn.grafika.FrameControlCallback
import com.lqk.videolearn.grafika.FramePlayer
import java.io.File
import java.io.IOException

class FramePlayerActivity : Activity(), View.OnClickListener, FramePlayer.PlayerFeedback {

    private var mTextureView: TextureView? = null

    private var mPlayTask: FramePlayer.PlayTask? = null

    private var mFps: TextView? = null

    private var mShowStopLabel = false


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_player_frame)
        mTextureView = findViewById(R.id.frame_player)
        findViewById<Button>(R.id.play).setOnClickListener(this)
        mFps = findViewById(R.id.fps)

        updateControls()
    }

    override fun onClick(v: View) = when (v.id) {
        R.id.play -> {
            clickPlayStop()
        }
        else -> {
        }
    }

    /**
     * 播放与暂停按钮
     */
    private fun clickPlayStop() {
        if (mShowStopLabel){
            stopPlayback()
        }else{
            if (mPlayTask != null) {
                return
            }

            val st: SurfaceTexture = mTextureView!!.surfaceTexture
            val surface = Surface(st)

            val framePlayer : FramePlayer
            try {
                val callback = FrameControlCallback()

                val infoCallback = FramePlayer.InfoCallback{
                    runOnUiThread{
                        mFps!!.text = String.format("FPS: %d", it)
                    }
                }
                framePlayer = FramePlayer(File("sdcard/v1080.mp4"),
                        surface, callback, infoCallback)
                framePlayer.setLoop(true)

                if (framePlayer.frameRate != 0){
                    callback.setFps(framePlayer.frameRate)
                }else{
                    callback.setFps(30)
                }

            } catch (e: IOException) {
                surface.release()
                return
            }

            adjustAspectRatio(framePlayer.width, framePlayer.height)
            mPlayTask = FramePlayer.PlayTask(framePlayer, this)

            mShowStopLabel = true
            updateControls()
            mPlayTask!!.execute()
        }
    }

    /**
     * 调整TextureView的显示矩阵
     */
    private fun adjustAspectRatio(videoWidth: Int, videoHeight: Int) {
        val surfaceWidth: Int = mTextureView!!.width
        val surfaceHeight: Int = mTextureView!!.height
        val aspectRatio: Double = videoHeight.toDouble() / videoWidth

        val newWidth: Int
        val newHeight: Int
        when (surfaceHeight > (videoWidth * aspectRatio).toInt()) {
            true -> {
                // limited by narrow width; restrict height
                newWidth = surfaceWidth
                newHeight = (surfaceWidth * aspectRatio).toInt()
            }
            false -> {
                // limited by short height; restrict width
                newWidth = (surfaceHeight / aspectRatio).toInt()
                newHeight = surfaceHeight
            }
        }
        val xOff = (surfaceWidth - newWidth) / 2
        val yOff = (surfaceHeight - newHeight) / 2

        val txForm = Matrix()
        mTextureView!!.getTransform(txForm)
        txForm.setScale(newWidth.toFloat() / surfaceWidth, newHeight.toFloat() / surfaceHeight)
//        txForm.postRotate(10F)
        txForm.postTranslate(xOff.toFloat(), yOff.toFloat())
        mTextureView!!.setTransform(txForm)

    }

    private fun stopPlayback() {
        if (mPlayTask != null) {
            mPlayTask!!.requestStop()
        }
    }

    override fun playbackStopped() {
        mShowStopLabel = false
        mPlayTask = null
        updateControls()
    }

    private fun updateControls() {
        val play = findViewById<View>(R.id.play) as Button
        if (mShowStopLabel) {
            play.setText(R.string.stop_button_text)
        } else {
            play.setText(R.string.play_button_text)
        }
    }

    override fun onPause() {
        // 确保暂停后 不再发送帧，因为view破坏了
        if (mPlayTask != null) {
            stopPlayback()
            mPlayTask!!.waitForStop()
        }

        super.onPause()
    }
}